k-distribution/src/main/scripts/lib/pyk/pyk/kastManip.py

```diff
@@ -212,28 +266,59 @@ def addOccurance(k):
     return occurances
 
 
-def countVarOccurances(kast, numOccurances=None):
-    """Count the number of occurances of each variable in a proof.
+def extract_lhs(term: KInner) -> KInner:
+    return traverseTopDown(term, if_ktype(KRewrite, lambda rw: rw.lhs))
 
-    -   Input: Kast term.
-    -   Output: Map of variable names to their number of occurances.
-    """
-    numOccurances = {} if numOccurances is None else numOccurances
 
-    def _getNumOccurances(_kast):
-        if type(_kast) is KVariable:
-            vName = _kast.name
-            if vName in numOccurances:
-                numOccurances[vName] += 1
-            else:
-                numOccurances[vName] = 1
+def extract_rhs(term: KInner) -> KInner:
+    return traverseTopDown(term, if_ktype(KRewrite, lambda rw: rw.rhs))
+
+
+def count_vars(term: KInner) -> Counter:
+    counter: Counter = Counter()
+
+    def count(term: KInner) -> None:
+        if type(term) is KVariable:
+            counter[term.name] += 1
 
-    collectBottomUp(kast, _getNumOccurances)
-    return numOccurances
+    collectBottomUp(term, count)
+    return counter
```

Comment:
Please include a test for `count_vars`. `collectBottomUp` has given me weird problems in the past and I'd like to make sure that we are locking in the correct behavior.

Reply:
Tests added.